parser/auth/sm3.go

```diff
@@ -0,0 +1,384 @@
+
+	// 1, 2, 3
+	bufA := bytes.NewBuffer(make([]byte, 0, 4096))
+	bufA.Write([]byte(plaintext))
+	bufA.Write(salt)
+
+	// 4, 5, 6, 7, 8
+	bufB := bytes.NewBuffer(make([]byte, 0, 4096))
+	bufB.Write([]byte(plaintext))
+	bufB.Write(salt)
+	bufB.Write([]byte(plaintext))
+	sumB := SM3(bufB.Bytes())
+	bufB.Reset()
+
+	// 9, 10
+	var i int
+	for i = len(plaintext); i > MIXCHARS; i -= MIXCHARS {
+		bufA.Write(sumB[:MIXCHARS])
+	}
+	bufA.Write(sumB[:i])
+
+	// 11
+	for i = len(plaintext); i > 0; i >>= 1 {
+		if i%2 == 0 {
+			bufA.Write([]byte(plaintext))
+		} else {
+			bufA.Write(sumB[:])
+		}
+	}
+
+	// 12
+	sumA := SM3(bufA.Bytes())
+	bufA.Reset()
+
+	// 13, 14, 15
+	bufDP := bufA
+	for range []byte(plaintext) {
+		bufDP.Write([]byte(plaintext))
+	}
+	sumDP := SM3(bufDP.Bytes())
+	bufDP.Reset()
+
+	// 16
+	p := make([]byte, 0, sha256.Size)
+	for i = len(plaintext); i > 0; i -= MIXCHARS {
+		if i > MIXCHARS {
+			p = append(p, sumDP[:]...)
+		} else {
+			p = append(p, sumDP[0:i]...)
+		}
+	}
+
+	// 17, 18, 19
+	bufDS := bufA
+	for i = 0; i < 16+int(sumA[0]); i++ {
+		bufDS.Write(salt)
+	}
+	sumDS := SM3(bufDS.Bytes())
+	bufDS.Reset()
+
+	// 20
+	s := make([]byte, 0, 32)
+	for i = len(salt); i > 0; i -= MIXCHARS {
+		if i > MIXCHARS {
+			s = append(s, sumDS[:]...)
+		} else {
+			s = append(s, sumDS[0:i]...)
+		}
+	}
+
+	// 21
+	bufC := bufA
+	var sumC []byte
+	for i = 0; i < iterations; i++ {
+		bufC.Reset()
+		if i&1 != 0 {
+			bufC.Write(p)
+		} else {
+			bufC.Write(sumA[:])
+		}
+		if i%3 != 0 {
+			bufC.Write(s)
+		}
+		if i%7 != 0 {
+			bufC.Write(p)
+		}
+		if i&1 != 0 {
+			bufC.Write(sumA[:])
+		} else {
+			bufC.Write(p)
+		}
+		sumC = SM3(bufC.Bytes())
+		sumA = sumC
+	}
+
+	// 22
+	buf := bytes.NewBuffer(make([]byte, 0, 100))
+	buf.Write([]byte{'$', 'B', '$'})
+	rounds := fmt.Sprintf("%03d", iterations/ITERATION_MULTIPLIER)
+	buf.Write([]byte(rounds))
+	buf.Write([]byte{'$'})
+	buf.Write(salt)
+
+	b64From24bit([]byte{sumC[0], sumC[10], sumC[20]}, 4, buf)
+	b64From24bit([]byte{sumC[21], sumC[1], sumC[11]}, 4, buf)
+	b64From24bit([]byte{sumC[12], sumC[22], sumC[2]}, 4, buf)
+	b64From24bit([]byte{sumC[3], sumC[13], sumC[23]}, 4, buf)
+	b64From24bit([]byte{sumC[24], sumC[4], sumC[14]}, 4, buf)
+	b64From24bit([]byte{sumC[15], sumC[25], sumC[5]}, 4, buf)
+	b64From24bit([]byte{sumC[6], sumC[16], sumC[26]}, 4, buf)
+	b64From24bit([]byte{sumC[27], sumC[7], sumC[17]}, 4, buf)
+	b64From24bit([]byte{sumC[18], sumC[28], sumC[8]}, 4, buf)
+	b64From24bit([]byte{sumC[9], sumC[19], sumC[29]}, 4, buf)
+	b64From24bit([]byte{0, sumC[31], sumC[30]}, 3, buf)
+
+	return buf.String()
+}
+
+// CheckSM3Password checks if a sm3 authentication string matches a password
+func CheckSM3Password(pwhash []byte, password string) (bool, error) {
+	pwhashParts := bytes.Split(pwhash, []byte("$"))
+	if len(pwhashParts) != 4 {
+		return false, errors.New("failed to decode hash parts")
+	}
+
+	hashType := string(pwhashParts[1])
+	if hashType != "B" {
```

Comment:
Type `A` is `SHA256` ([source](https://dev.mysql.com/doc/dev/mysql-server/latest/classsha2__password_1_1Caching__sha2__password.html#a557d477537771825a0fad8fe0aa2446f)). Why did we pick `B` here? Could this lead to problems if MySQL ever started to use this for SHA512 or something else?

Reply:
Since our plugin has been renamed to `tidb_sm3_password`, it would not conflict with MySQL's current and future authentication.